### DISCUSSION TITLE
Packaging my text-user-interface and its requirements.txt libraries into a "1-click" executable

### DISCUSSION THREAD

**Discussion by ishaan-ghosh** [Textualize/textual#4512] [Q&A] [answered]
Hi Guys! I have built a pretty simple text-user-interface with textual for a third-party command line interface library I am utilizing. My current objective is to package the TUI as a binary/executable, I want the executable to be "1-click", meaning I don't need the end-user to install any libraries (the CLI or textual), or even python itself honestly (even though that should be very uncommon). 

The basic project hierarchy is like so:
```
TUI
├── requirements.txt
├── application
│   ├── main.py
│   ├── app
│   │   ├── frontend.py
│   │   ├── style.tcss
│   ├── back
│   │   ├── backend.py
``` 
The CLI I am building the TUI for is a part of the PyPi repository, and is including in my TUI requirements.txt, so it should be easy to add to the packaged executable. 

How would I go about doing something like this? I have already attempted using PyInstaller using the following command:

```
pyinstaller --onefile --add-data 'application/app/style.tcss:application/app/' --name "TUI v0.1.0 (MacOS)" application/main.py
``` 

However, it doesn't seem to package the CLI with the TUI. When I attempt to run the TUI it hits my internal error check for whether the TUI can be called, and errors out. Does anybody have prior experience packaging a small textual project with additional python library requirements like this? If so, I would greatly appreciate some advice on whether a different packaging library worked better for you, or if I may be doing something wrong with PyInstaller.

Any advice is much appreciated!

**Comment by TomJGooding** [comment]
Today was my first time trying  `pyinstaller` so I'm probably not going to be much help, but I'm not sure there's really enough detail here for others to offer any advice.

> When I attempt to run the TUI it hits my internal error check for whether the TUI can be called, and errors out. 

What exactly is that code and the traceback?

**Comment by jamesreverett** [comment]
I have used pyinstaller for other applications (Tkinter UIs) successfully and am now building an app using pyinstaller with Textualize (for the first time). 

Don't know if this will be helpful or prompt you to discover your specific solution but ran into an issue today where pyinstaller was missing a Textualize module. Generally pyinstaller is pretty good at discovering all the dependencies and if not there is a mechanism to force things. 

The runtime issue Traceback was:
```
in compose:115
 
in __getattr__:106
  
in import_module:126
in _gcd_import:1204
----- locals -----
│ │   level = 1
│ │    name = 'textual.widgets._tab_pane'
│ │ package = 'textual.widgets'

│ in _find_and_load:1176
----- locals -----
│ │ import_ = <function _gcd_import at 0x000002ABCC903D80>
│ │  module = <object object at 0x000002ABCC934050>
│ │    name = 'textual.widgets._tab_pane'
│
│ in _find_and_load_unlocked:1140
----- locals -----
│ │         child = '_tab_pane'  
│ │       import_ = <function _gcd_import at 0x000002ABCC903D80> 
│ │          name = 'textual.widgets._tab_pane'
│ │        parent = 'textual.widgets'
│ │ parent_module = <module 'textual.widgets' from
│ │                 'C:\\[MY APP]\\_internal\\textual\\widgets\\__init__.pyc'>    
│ │   parent_spec = ModuleSpec(name='textual.widgets', loader=<pyimod02_importers.PyiFrozenImporter object at 0x000002ABCCB1CF10>,  
│ │                 origin='C:\\[MY APP]\\_internal\\textual\\widgets\\__init__.pyc',  
│ │                 submodule_search_locations=['C:\\[MY APP]\\_internal\\textual\\widgets']) 
│ │          path = ['C:\\[MY APP]\\_internal\\t'+14]
│ │          spec = None
│
ModuleNotFoundError: No module named 'textual.widgets._tab_pane'
```

To resolve this, I added the following to my pyinstaller Analysis():

`hiddenimports = ['textual.widgets._tab_pane']`

Looking at the latest (0.67.1) textualize source it appears that "_tab_pane" is not referenced directly anywhere but perhaps I'm missing something. Currently, I'm only using a handful of widgets and containers so did not do a thorough search on other modules.

Btw, I'm using pyinstaller 6.3.0 in folder mode (not onefile) and a .spec file.
Also using poetry and invoke for build infrastructure.

Hope that helps.

**Comment by KRRT7** [comment, marked as answer]
Another option here would be Nuitka, which can accomplish the same thing using it's `--onefile` mode, it should work out of the box, if it doesn't, reach out to me and i can help :)

**Reply by Devanshi-Crypto** [reply]
Thanks man, this worked for me as well

**Reply by moonblind** [reply]
hi -- nuitka's `onefile` mode works for generating a `.bin` for me on macOS, but does not seem to work for making an `.app` which a user can click to execute. i'm assuming this is WAI, since i'm not sure how an executable would open a terminal in the first place, but i just wanted to confirm that?

**Reply by KRRT7** [reply]
hi @moonblind I don't have much experience with packaging on macOS but I believe on macOS you want `nuitka --mode=app main.py`  rather than `nuitka --onefile main.py`

**Reply by moonblind** [reply]
hanks for the reply @KRRT7  -- I tried compiling both with `--mode=app` and `--macos-create-app-bundle`, but neither created an `.app` that would open a textual TUI when clicked; i only seem to be able to launch a `.bin` via the terminal otherwise. do you know if there is any prior art for someone packaging a textual project on mac via nuitka?

**Reply by KRRT7** [reply]
@moonblind no worries, mind joining the Textual discord? I'll look into it today and see what I come up with, but it'll be easier to communicate there.

**Comment by dontascii** [comment]
A little late here, but I've had great success using [Shiv](https://shiv.readthedocs.io/en/latest/), from the LinkedIn devs.  It works just like zip apps do,  ([PEP 441](http://legacy.python.org/dev/peps/pep-0441/)) but it includes all your dependencies, like [PEX](https://github.com/pantsbuild/pex). It performs better than PEX, however and has some neat options such as `reproducible`, to generate a reproducible zipapp and `no-modify`, which will detect any changes to the unzipped source files and throw an exception.
